### PR TITLE
Open a previewEditor while navigating files through QuickOpen

### PR DESCRIPTION
**Pull request by XVincentX**
The following PR will solve #12892 - allowing to have a preview editor while navigating into files from a QuickOpen menu.

### CODE REVIEW

File: src/vs/workbench/browser/parts/editor/editorPart.ts [hunk 1]
```diff
@@ -404,7 +404,7 @@ export class EditorPart extends Part implements IEditorPart, IEditorGroupService
 			}
 
 			// Focus (unless prevented)
-			const focus = !options || !options.preserveFocus;
+			const focus = !options || options.preserveFocus;
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorPart.ts]: This cannot be changed...

Comment (XVincentX) [inline src/vs/workbench/browser/parts/editor/editorPart.ts]: Ok, I understand that.
If the `options` object is undefined it's going to focus the thing anyway - how should we handle this situation?

File: src/vs/workbench/browser/quickopen.ts [hunk 2]
```diff
@@ -247,20 +247,26 @@ export class EditorQuickOpenEntry extends QuickOpenEntry implements IEditorQuick
...
+
+		let input = this.getInput();
+		let options = this.getOptions();
 
-			return true;
+		const isPreview = mode === Mode.PREVIEW;
+
+		if (input instanceof EditorInput) {
+			if (isPreview) {
+				options.preserveFocus = false;
```
Comment (bpasero) [inline src/vs/workbench/browser/quickopen.ts]: flip this

File: src/vs/workbench/common/editor.ts [hunk 3]
```diff
@@ -532,7 +532,7 @@ export class TextEditorOptions extends EditorOptions {
 	public static from(input: IResourceInput): TextEditorOptions {
 		let options: TextEditorOptions = null;
 		if (input && input.options) {
-			if (input.options.selection || input.options.forceOpen || input.options.revealIfVisible || input.options.preserveFocus || input.options.pinned || input.options.inactive || typeof input.options.index === 'number') {
+			if (input.options.selection || input.options.forceOpen || input.options.revealIfVisible || !input.options.preserveFocus || input.options.pinned || input.options.inactive || typeof input.options.index === 'number') {
```
Comment (bpasero) [inline src/vs/workbench/common/editor.ts]: all of this cannot be changed...

File: src/vs/workbench/browser/parts/editor/editorPart.ts (hunk 1 above)
Comment (bpasero) [inline src/vs/workbench/browser/parts/editor/editorPart.ts]: that is the correct behaviour. keeping focus should be the exception so please leave it like that

**Comment by XVincentX** [comment]
@bpasero Cool - I misinterpreted the meaning of `preserveFocus`
I've flipped the behaviours and now it works correctly!

Thanks for the patience - I'll keep going and see what I can find out for the history entry!

**Comment by bpasero** [comment]
@XVincentX I am not 100% convinced that this should be an editor option. It seems very specific to the history management so maybe it would be better to just add a method to the history service to temporarily ignore editor changes while you navigate through the results. 

There are 2 ugly cases that we still need to figure out: If you check for listeners of `onEditorOpening` you can find:
* the remove from editor history action that waits for a selection being made in quick open to do something (https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/browser/parts/editor/editorActions.ts#L1168)
* the "compare active file with" action that does a similar trick (https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/parts/files/browser/fileActions.ts#L1248)

Obviously those two would not work anymore if we start to preview editors while navigating. i wonder if another solution should be added to avoid the `onEditorOpening` thingy.

**Comment by bpasero** [comment]
I think the right approach to solve this issue is to change the `IQuickOpenService.show()` method to allow to be run in a mode where the result is returned, e.g. by passing in a function that should be called when an element is selected. If this is provided, the default action in the `run` method would not be called. That needs quite some adoption though across quick open providers...

**Comment by XVincentX** [comment]
@bpasero I see.
I am totally willing to contribute! However if we do not really have a concrete action plan (agreed by you and probably your team mates) I have no idea how to start and make sure I do not come up with wrong solutions.

**Comment by bpasero** [comment]
@XVincentX I pushed a change to get rid of onEditorOpening because I felt it was very ugly. This should make your PR simpler now. I think you should push for a method in history service to simply ignore pushing to the stack and try to leverage that 👍

**Comment by XVincentX** [comment]
@bpasero Could you link the commit/change you made? So I can understand what happened in particular.

**Comment by bpasero** [comment]
@XVincentX here: https://github.com/Microsoft/vscode/commit/dfd3f574fc65d5aad06c8a2f877e82f4fb82d8bf

This event would always fire when an editor is about to open and it would interfere with your contribution to open editors while navigating over results in the picker. I was never fond of that event and now decided to get rid of it. This should make your life easier.

**Comment by XVincentX** [comment]
Thanks! Let’s see if I can come up with something.

V.

**Comment by XVincentX** [comment]
@bpasero Unfortunately I will not be able to work on this during this week as I'll be out for the entire week. Please bear with me, hopefully I should be able to get the hands on next week.

**Comment by bpasero** [comment]
@XVincentX np at all 👍

**Comment by bpasero** [comment]
Closing in favour of https://github.com/Microsoft/vscode/pull/13909

**Pull request closed without merging**